packages/stateful/actions/actions/AuthzExec.tsx

```diff
@@ -0,0 +1,214 @@
+import type {
+  MsgBeginRedelegate,
+  MsgDelegate,
+  MsgUndelegate,
+} from 'cosmjs-types/cosmos/staking/v1beta1/tx'
+import { useCallback, useMemo } from 'react'
+
+import { validatorsSelector } from '@dao-dao/state/recoil'
+import {
+  ActionCardLoader,
+  LockWithKeyEmoji,
+  useCachedLoadable,
+} from '@dao-dao/stateless'
+import {
+  ActionComponent,
+  ActionMaker,
+  CoreActionKey,
+  UseDecodedCosmosMsg,
+  UseDefaults,
+  UseTransformToCosmos,
+} from '@dao-dao/types/actions'
+import {
+  NATIVE_DENOM,
+  loadableToLoadingData,
+  makeRawProtobufMsg,
+  makeStargateMessage,
+} from '@dao-dao/utils'
+
+import { ProfileDisplay, SuspenseLoader } from '../../components'
+import { AuthzExecComponent as StatelessAuthzComponent } from '../components'
+import { useActionOptions } from '../react'
+
+export enum AuthzExecActionTypes {
+  Delegate = '/cosmos.staking.v1beta1.MsgDelegate',
+  Undelegate = '/cosmos.staking.v1beta1.MsgUndelegate',
+  Redelegate = '/cosmos.staking.v1beta1.MsgBeginRedelegate',
+  ClaimRewards = '/cosmos.distribution.v1beta1.MsgWithdrawDelegatorReward',
+  Custom = 'custom',
+}
+
+interface AuthzExecData {
+  authzExecActionType: AuthzExecActionTypes
+  claimRewards: MsgWithdrawDelegatorReward
+  custom: string
+  delegate: MsgDelegate
+  redelegate: MsgBeginRedelegate
+  undelegate: MsgUndelegate
+}
+
+const useDefaults: UseDefaults<AuthzExecData> = () => ({
+  authzExecActionType: AuthzExecActionTypes.Delegate,
+  grantee: '',
+  claimRewards: {
+    delegatorAddress: '',
+    validatorAddress: '',
+  },
+  custom: '[]',
+  delegate: {
+    amount: { denom: NATIVE_DENOM, amount: '0' },
+    delegatorAddress: '',
+    validatorAddress: '',
+  },
+  undelegate: {
+    amount: {
+      denom: NATIVE_DENOM,
+      amount: '0',
+    },
+    delegatorAddress: '',
+    validatorAddress: '',
+  },
+  redelegate: {
+    delegatorAddress: '',
+    validatorSrcAddress: '',
+    validatorDstAddress: '',
+    amount: {
+      denom: NATIVE_DENOM,
+      amount: '0',
+    },
+  },
+})
+
+const Component: ActionComponent = (props) => {
+  const { chainId } = useActionOptions()
+
+  const loadingValidators = loadableToLoadingData(
+    useCachedLoadable(
+      validatorsSelector({
+        chainId,
+      })
+    ),
+    []
+  )
+
+  return (
+    <SuspenseLoader fallback={<ActionCardLoader />}>
+      <StatelessAuthzComponent
+        {...props}
+        options={{
+          ProfileDisplay: ProfileDisplay,
+          validators: loadingValidators.loading ? [] : loadingValidators.data,
+        }}
+      />
+    </SuspenseLoader>
+  )
+}
+
+const useDecodedCosmosMsg: UseDecodedCosmosMsg<AuthzExecData> = (
+  msg: Record<string, any>
+) =>
+  useMemo(
+    () =>
+      'stargate' in msg &&
+      msg.stargate.type_url === '/cosmos.authz.v1beta1.MsgExec' &&
+      msg.stargate.value
+        ? {
+            match: true,
+            data: msg as AuthzExecData,
```

Comment:
`msg` is going to look like:

```
{
  stargate: {
    type_url: '/cosmos.authz.v1beta1.MsgExec',
    value: { ... }
  }
}
```

you're going to have to check that `msg.stargate.value.msgs` has only 1 message, since `useTransformToCosmos` below seems to create stargate messages with only 1 message each, and use the `typeUrl` of that first message to set `data.authzExecActionType` as well as one of the `data.delegate`, `data.redelegate`, `data.undelegate`, `data.claimRewards`, and `data.custom` fields accordingly. this `data` field has to match the schema of `AuthzExecData`, but right now you're just casting it.